pkg/server/plugin/nodeattestor/httpchallenge/httpchallenge_test.go

```diff
@@ -0,0 +1,423 @@
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.tofu {
+				common_httpchallenge.DefaultRandReader = strings.NewReader("bad123456789abcdefghijklmnopqrstuvwxyz")
+			} else {
+				common_httpchallenge.DefaultRandReader = strings.NewReader("123456789abcdefghijklmnopqrstuvwxyz")
+			}
+			plugin := loadPlugin(t, tt.hclConf, !tt.tofu)
+			result, err := plugin.Attest(context.Background(), tt.payload, tt.challengeFn)
+			require.Contains(t, err.Error(), tt.expErr)
+			require.Nil(t, result)
+		})
+	}
+}
+
+func TestAttestSucceeds(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	    if r.URL.Path != "/.well-known/spiffe/nodeattestor/http_challenge/default/challenge" {
+	        t.Errorf("Expected to request '/.well-known/spiffe/nodeattestor/http_challenge/default/challenge', got: %s", r.URL.Path)
+	    }
+	    w.WriteHeader(http.StatusOK)
+	    w.Write([]byte(`MTIzNDU2Nzg5YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnc=`))
+	}))
+	defer server.Close()
+
+	u, _ := url.Parse(server.URL)
+	_, port, _ := net.SplitHostPort(u.Host)
+	oldDialContext := http.DefaultTransport.(*http.Transport).DialContext
+	dialContext := func(ctx context.Context, network, addr string) (net.Conn, error) {
+		if addr == "foo:80" {
+			addr = fmt.Sprintf("127.0.0.1:%s", port)
+		}
+		return oldDialContext(ctx, network, addr)
+	}
+	http.DefaultTransport.(*http.Transport).DialContext = dialContext
+
+	challengeFnNil := func(ctx context.Context, challenge []byte) ([]byte, error) {
+		return nil, nil
+	}
+
+	tests := []struct {
+		name        string
+		hclConf     string
+		payload     []byte
+		challengeFn func(ctx context.Context, challenge []byte) ([]byte, error)
+		expectedAgentID   string
+		expectedSelectors []*common.Selector
+		tofu        bool
+	}{
+		{
+			name:        "Attest succeeds for defaults",
+			hclConf:     "",
+			tofu: true,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      80,
+			}),
+			expectedAgentID: "spiffe://example.org/spire/agent/http_challenge/foo",
+			expectedSelectors: []*common.Selector{
+				{
+					Type:  "http_challenge",
+					Value: "hostname:foo",
+				},
+			},
+		},
+		{
+			name:        "Attest succeeds for reattest without tofu",
+			hclConf:     "tofu = false\nallow_non_root_ports = false",
+			tofu: false,
+			challengeFn: challengeFnNil,
+			payload: marshalPayload(t, &common_httpchallenge.AttestationData{
+				HostName:  "foo",
+				AgentName: "default",
+				Port:      80,
+			}),
+			expectedAgentID: "spiffe://example.org/spire/agent/http_challenge/foo",
+			expectedSelectors: []*common.Selector{
+				{
+					Type:  "http_challenge",
+					Value: "hostname:foo",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			common_httpchallenge.DefaultRandReader = strings.NewReader("123456789abcdefghijklmnopqrstuvwxyz")
```

Comment:
If possible, it's better to avoid mutating package level variables from tests. Doing so inhibits parallelizing test execution in the future.

Instead of trying to control the generated nonce, can the closure you pass in for the challenge response function update a variable that controls what is returned from the http server?